cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/S3BatchSource.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.template.etl.batch.source;
+
+import co.cask.cdap.api.annotation.Description;
+import co.cask.cdap.api.annotation.Name;
+import co.cask.cdap.api.annotation.Plugin;
+import co.cask.cdap.api.templates.plugins.PluginConfig;
+import co.cask.cdap.template.etl.api.batch.BatchSource;
+
+import javax.annotation.Nullable;
+
+/**
+ * A {@link BatchSource} that reads from Amazon S3.
+ */
+@Plugin(type = "source")
+@Name("S3BatchSource")
+@Description("Batch source for Amazon S3")
+public class S3BatchSource extends FileBatchSource {
+  private static final String ACCESS_ID_DESCRIPTION = "Access ID of Amazon S3 instance to connect to.";
+  private static final String ACCESS_KEY_DESCRIPTION = "Access Key of Amazon S3 instance to connect to.";
+
+  private final S3BatchConfig config;
+
+  public S3BatchSource(S3BatchConfig config) {
```

Comment:
so interestingly enough, not only must the constructor take a single argument in the type of the config field, but the class must also contain a field of the config type. Found this out because the following constructor does not get called
public TimePartitionedFileSetDatasetAvroSink(TPFSAvroSinkConfig config) {
    super(new TPFSSinkConfig(config.name, config.basePath));
    this.schema = config.schema;
  }
unless the class has a field private TPFSAvroSinkConfig config. Is this intended behavior?